beep/structure/base.py

```diff
@@ -0,0 +1,1373 @@
+            "internal_resistance": "last",
+        }
+
+        if self.raw_data is not None and "temperature" in self.raw_data.columns:
+            self._aggregation["temperature"] = ["max", "mean", "min"]
+
+            self._summary_cols = [
+                "cycle_index",
+                "discharge_capacity",
+                "charge_capacity",
+                "discharge_energy",
+                "charge_energy",
+                "dc_internal_resistance",
+                "temperature_maximum",
+                "temperature_average",
+                "temperature_minimum",
+                "date_time_iso",
+            ]
+        else:
+            self._summary_cols = [
+                "cycle_index",
+                "discharge_capacity",
+                "charge_capacity",
+                "discharge_energy",
+                "charge_energy",
+                "dc_internal_resistance",
+                "date_time_iso",
+            ]
+
+        # Ensure this step of the aggregation is placed last
+        self._aggregation["date_time_iso"] = "first"
+
+        self._diag_aggregation = copy.deepcopy(self._aggregation)
+        self._diag_aggregation.pop("internal_resistance")
+        self._diag_summary_cols = copy.deepcopy(self._summary_cols)
+        self._diag_summary_cols.pop(5)  # internal_resistance
+
+    @classmethod
+    @abc.abstractmethod
+    def from_file(cls, path, *args, **kwargs):
+        """Go from a raw cycler output data file (or files) to a BEEPDatapath.
+
+        Must be implemented for the BEEPDatapath child to be valid.
+
+        Args:
+            path (str, Pathlike): The path to the raw data file.
+
+        Returns:
+            (BEEPDatapath): A child class of BEEPDatapath.
+
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def from_json_file(cls, filename):
+        """Load a structured run previously saved to file.
+
+        Loads a BEEPDatapath or (legacy) ProcessedCyclerRun structured object from json.
+
+        Can be used in combination with files serialized with BEEPDatapath.to_json_file.
+
+        Args:
+            filename (str, Pathlike): a json file from a structured run, serialzed with to_json_file.
+
+        Returns:
+            None
+        """
+        with open(filename, "r") as f:
+            d = json.load(f)
+
+        # Add this structured file path to the paths dict
+        paths = d.get("paths", {})
+        paths["structured"] = os.path.abspath(filename)
+        d["paths"] = paths
+
+        return cls.from_dict(d)
+
+    def to_json_file(self, filename):
+        """Save a BEEPDatapath to disk as a json.
+
+        Not named from_json to avoid conflict with MSONable.from_json(*)
+
+        Args:
+            filename (str, Pathlike): The filename to save the file to.
+
+        Returns:
+            None
+        """
+        with open(filename, "w") as f:
+            json.dump(self.as_dict(), f)
+
+    @StructuringDecorators.must_not_be_legacy
```

Comment:
I like the use of the structure decorators!

Reply:
Thanks! :)